graylog2-server/src/main/java/org/graylog2/indexer/messages/SerializationMemoizingMessage.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.indexer.messages;
+
+import com.codahale.metrics.Meter;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import jakarta.annotation.Nonnull;
+import org.graylog2.indexer.IndexSet;
+import org.graylog2.plugin.Message;
+import org.graylog2.plugin.utilities.ratelimitedlog.RateLimitedLogFactory;
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.ref.SoftReference;
+import java.time.Duration;
+import java.util.Map;
+import java.util.Set;
+
+public class SerializationMemoizingMessage implements ImmutableMessage {
+    private static final Logger LOG = LoggerFactory.getLogger(SerializationMemoizingMessage.class);
+    private static final Logger RATE_LIMITED_LOG = RateLimitedLogFactory.createRateLimitedLog(
+            LOG, 1, Duration.ofMinutes(1));
+
+    private final Message delegate;
+
+    private volatile SoftReference<CacheEntry> lastSerializationResult;
+
+    private record CacheEntry(byte[] serializedBytes,
+                              ObjectMapper objectMapper,
+                              Meter invalidTimeStampMeter) {}
+
+    public SerializationMemoizingMessage(Message delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public synchronized byte[] serialize(SerializationContext context)
+            throws JsonProcessingException {
+
+        final var objectMapper = context.objectMapper();
+        final var invalidTimestampMeter = context.invalidTimestampMeter();
+
+        CacheEntry cachedEntry = null;
+        if (lastSerializationResult != null) {
+            cachedEntry = lastSerializationResult.get();
+            if (cachedEntry == null) {
+                RATE_LIMITED_LOG.warn("The JVM cleared a cached serialized message because of memory pressure. " +
+                        "This has a performance impact. Please adjust the memory configuration to assign more " +
+                        "heap memory to the JVM.");
+            }
+        }
+
+        if (cachedEntry == null || !cachedEntry.objectMapper().equals(objectMapper)) {
+            final var tsMeter = new Meter();
+            final var serializedBytes = delegate.serialize(new DefaultSerializationContext(objectMapper, tsMeter));
```

Comment:
Why are we using a new context and meter here?

Reply:
If we re-use the provided meter, then invalid timestamps will be counted only on the first serialization. I thought it would be counter-intuitive that if you call `#serialize` and provide a meter for invalid timestamps, you won't know if they are counted.

Reply:
Got it, thanks! I was focusing too much on the one branch in the code and missed that we take the meter from the cache entry to actually mark the meter from the context.